brave-core/src/main/java/com/github/kristofa/brave/internal/Propagation.java

```diff
@@ -0,0 +1,61 @@
+package com.github.kristofa.brave.internal;
+
+import com.github.kristofa.brave.SpanId;
+import com.github.kristofa.brave.TraceData;
+
+/**
+ * Injects and extracts {@link SpanId span identifiers} sent in-band across process boundaries.
+ *
+ * <p>Implementations usually wrap Http, Messaging or RPC headers.
+ *
+ * @param <K> Usually, but not always a String
+ */
+public interface Propagation<K> {
+
+  interface Factory {
+    Factory B3 = new Factory() {
+      @Override public <K> Propagation<K> create(KeyFactory<K> keyFactory) {
+        return B3Propagation.create(keyFactory);
+      }
+    };
+
+    <K> Propagation<K> create(KeyFactory<K> keyFactory);
+  }
+
+  /** Creates keys for use in propagated contexts */
+  interface KeyFactory<K> {
+    KeyFactory<String> STRING = new KeyFactory<String>() {
+      @Override public String create(String name) {
+        return name;
+      }
+    };
+
+    K create(String name);
+  }
+
+  /** Replaces a propagated key with the given value */
+  interface Setter<K> {
+    void put(K key, String value);
+  }
+
+  /** Gets the first value of the given key or returns null */
+  interface Getter<K> {
+    @Nullable String get(K key);
+  }
+
+  /**
+   * Calls the setter for each key to send downstream.
+   *
+   * @param spanId possibly null span identifiers. Null means unsampled.
+   * @param setter invoked for each propagation key to add.
+   */
+  void injectSpanId(@Nullable SpanId spanId, Setter<K> setter);
```

Comment:
these two methods obviate `ClientRequestAdapter.addSpanIdToRequest(SpanId)` and `ServerRequestAdapter.getTraceData()`. That makes the adapters' key focus parsing out-of-band data, as opposed to parsing out-of-band data and also hard-wiring propagation logic.